desc/basis.py

```diff
@@ -1413,6 +1413,424 @@ def zernike_radial_poly(r, l, m, dr=0, exact="auto"):
+            * factorial((k - abs(m)) / 2)
+        )
+
+    return R_nm
+
+
+@functools.partial(jit, static_argnums=3)
+def zernike_radial_optimized(r, l, m, dr=0):  # noqa: C901
+    """Radial part of zernike polynomials.
+
+    This is completely JITable version of zernike_radial_optimized
+    function.
+
+    Parameters
+    ----------
+    r : ndarray, shape(N,)
+        radial coordinates to evaluate basis
+    l : ndarray of int, shape(K,)
+        radial mode number(s)
+    m : ndarray of int, shape(K,)
+        azimuthal mode number(s)
+
+    Returns
+    -------
+    out : ndarray, shape(N,K)
+        basis function(s) evaluated at specified points
+
+    """
+    if dr > 4:
+        raise NotImplementedError(
+            "Analytic radial derivatives of Zernike polynomials for order>4 "
+            + "have not been implemented."
+        )
+
+    def body_inner(N, args):
+        r_jacobi, alpha, r, out, P_past, m, n = args
+        P_n1, P_n2 = P_past
+        # Calculate Jacobi polynomial for m,n pair
+        P_n = jacobi_poly_single(r_jacobi, N, alpha, 0, P_n1, P_n2)
+
+        # Find the index corresponding to the original array
+        index = jnp.where(
+            jnp.logical_and(m == alpha, n == N),
+            jnp.arange(m.size),
+            0,
+        )
+        # jnp.where() will result in 1D array with 1 value which
+        # is not 0. Sum it to find the index
+        idx = jnp.sum(index)
+
+        out = out.at[:, idx].set((-1) ** N * r**alpha * P_n)
+        P_n2 = jnp.where(N >= 2, P_n1, P_n2)
+        P_n1 = jnp.where(N >= 2, P_n, P_n1)
+        return (r_jacobi, alpha, r, out, (P_n1, P_n2), m, n)
+
+    def body_inner_d1(N, args):
+        r_jacobi, alpha, r, out, P_past, m, n = args
+        P_n1, P_n2, dP_n1, dP_n2 = P_past
+        # Calculate Jacobi polynomial for m,n pair
+        P_n = jacobi_poly_single(r_jacobi, N, alpha, 0, P_n1, P_n2)
+        dP_n = jacobi_poly_single(r_jacobi, N - dr, alpha + dr, dr, dP_n1, dP_n2)
+
+        # Find the index corresponding to the original array
+        index = jnp.where(
+            jnp.logical_and(m == alpha, n == N),
+            jnp.arange(m.size),
+            0,
+        )
+        # jnp.where() will result in 1D array with 1 value which
+        # is not 0. Sum it to find the index
+        idx = jnp.sum(index)
+
+        coef = gammaln(alpha + N + 1 + dr) - dr * jnp.log(2) - gammaln(alpha + N + 1)
+        coef = jnp.exp(coef)
+
+        result = (
+            alpha * r ** jnp.maximum(alpha - 1, 0) * P_n
+            - coef * 4 * r ** (alpha + 1) * dP_n
+        )
+        out = out.at[:, idx].set((-1) ** N * result)
+
+        P_n2 = jnp.where(N >= 2, P_n1, P_n2)
+        P_n1 = jnp.where(N >= 2, P_n, P_n1)
+        dP_n2 = jnp.where(N >= 3, dP_n1, dP_n2)
+        dP_n1 = jnp.where(N >= 3, dP_n, dP_n1)
+
+        return (r_jacobi, alpha, r, out, (P_n1, P_n2, dP_n1, dP_n2), m, n)
+
+    # USE DOT PRODUCT TO GENERALIZE DERIVATIVES
```

Comment:
Can you explain this a bit more?

Reply:
For the derivatives, we have derivative order+1 terms in result formula (i.e. a1*f + a2*df + a3*ddf + a4 * dddf, we can write this by calculatinf some A=[a1 a2 a3 a4] and F=[f df ddf dddf] and taking jnp.dot(A,F) ). I wanted to get some results first, before finalizing the algorithm.